msfdb

```diff
@@ -0,0 +1,816 @@
+
+  if @options[:ws_user].nil?
+    @msf_ws_user = ask_value("Initial MSF web service account username?", @msf_ws_user)
+  else
+    @msf_ws_user = @options[:ws_user]
+  end
+
+  # Write a default Rack config file for the web service
+  Dir.mkdir(@localconf) unless File.directory?(@localconf)
+  # TODO: free the REST API from all of these requirements
+  File.open(@ws_conf, 'w') do |f|
+    f.puts <<~EOF
+      # #{File.basename(@ws_conf)}
+      # created on: #{Time.now.utc}
+      require 'pathname'
+      require File.expand_path('./config/boot', '#{@framework}')
+      require 'metasploit/framework/parsed_options/remote_db'
+      require 'msf/core/db_manager/http/metasploit_api_app'
+
+      def require_environment!(parsed_options)
+        # RAILS_ENV must be set before requiring 'config/application.rb'
+        parsed_options.environment!
+        ARGV.replace(parsed_options.positional)
+
+        # allow other Rails::Applications to use this command
+        if !defined?(Rails) || Rails.application.nil?
+          # @see https://github.com/rails/rails/blob/v3.2.17/railties/lib/rails/commands.rb#L39-L40
+          require File.expand_path('./config/application', '#{@framework}')
+        end
+
+        # have to configure before requiring environment because
+        # config/environment.rb calls initialize! and the initializers will use
+        # the configuration from the parsed options.
+        parsed_options.configure(Rails.application)
+
+        Rails.application.require_environment!
+      end
+
+      parsed_options = Metasploit::Framework::ParsedOptions::RemoteDB.new
+      require_environment!(parsed_options)
+
+      run MetasploitApiApp
+    EOF
+  end
+  File.chmod(0640, @ws_conf)
+
+  if @options[:ssl] && ((!File.file?(@options[:ssl_key]) || !File.file?(@options[:ssl_cert])) ||
+      (@options[:ssl_key] == @ws_ssl_key_default && @options[:ssl_cert] == @ws_ssl_cert_default))
+    generate_web_service_ssl(key: @options[:ssl_key], cert: @options[:ssl_cert])
+  end
+
+  if !start_web_service
+    return
+  end
+
+  # wait until web service is online
+  retry_count = 0
+  is_online = web_service_online
+  while !is_online && retry_count < @options[:retry_max]
+    retry_count += 1
+    if @options[:debug]
+      puts "MSF web service doesn't appear to be online. Sleeping #{@options[:retry_delay]}s until check #{retry_count}/#{@options[:retry_max]}"
+    end
+    sleep(@options[:retry_delay])
+    is_online = web_service_online
+  end
+
+  if is_online
+    if add_web_service_user
+      output_web_service_information
+    end
+  else
+    puts "MSF web service does not appear to be online; aborting initialize."
+  end
+end
+
+def start_web_service
+  unless File.file?(@ws_conf)
+    puts "No MSF web service configuration found at #{@ws_conf}, not starting"
+    return false
+  end
+
+  # daemonize MSF web service
+  puts "Starting MSF web service"
+  if run_cmd("#{thin_cmd} start") == 0
+    puts "MSF web service started"
```

Comment:
This message is a little misleading. See the output below:

```
Starting MSF web service
MSF web service started
MSF web service does not appear to be online; aborting initialize.
```

It says that the web service has started, but it never successfully did. I had another web server running on the same port so it never started up, but the logging makes it seem like it did.

msf-ws.log:
```
/home/jbarnett/.rvm/gems/ruby-2.5.1@metasploit-framework/gems/eventmachine-1.2.7/lib/eventmachine.rb:531:in `start_tcp_server': no acceptor (port is in use or requires root privileges) (RuntimeError)
	from /home/jbarnett/.rvm/gems/ruby-2.5.1@metasploit-framework/gems/eventmachine-1.2.7/lib/eventmachine.rb:531:in `start_server'
	from /home/jbarnett/.rvm/gems/ruby-2.5.1@metasploit-framework/gems/thin-1.7.2/lib/thin/backends/tcp_server.rb:16:in `connect'
	from /home/jbarnett/.rvm/gems/ruby-2.5.1@metasploit-framework/gems/thin-1.7.2/lib/thin/backends/base.rb:63:in `block in start'
	from /home/jbarnett/.rvm/gems/ruby-2.5.1@metasploit-framework/gems/eventmachine-1.2.7/lib/eventmachine.rb:195:in `run_machine'
	from /home/jbarnett/.rvm/gems/ruby-2.5.1@metasploit-framework/gems/eventmachine-1.2.7/lib/eventmachine.rb:195:in `run'
	from /home/jbarnett/.rvm/gems/ruby-2.5.1@metasploit-framework/gems/thin-1.7.2/lib/thin/backends/base.rb:73:in `start'
	from /home/jbarnett/.rvm/gems/ruby-2.5.1@metasploit-framework/gems/thin-1.7.2/lib/thin/server.rb:162:in `start'
	from /home/jbarnett/.rvm/gems/ruby-2.5.1@metasploit-framework/gems/thin-1.7.2/lib/thin/controllers/controller.rb:87:in `start'
	from /home/jbarnett/.rvm/gems/ruby-2.5.1@metasploit-framework/gems/thin-1.7.2/lib/thin/runner.rb:203:in `run_command'
	from /home/jbarnett/.rvm/gems/ruby-2.5.1@metasploit-framework/gems/thin-1.7.2/lib/thin/runner.rb:159:in `run!'
	from /home/jbarnett/.rvm/gems/ruby-2.5.1@metasploit-framework/gems/thin-1.7.2/bin/thin:6:in `<top (required)>'
	from /home/jbarnett/.rvm/gems/ruby-2.5.1@metasploit-framework/bin/thin:23:in `load'
	from /home/jbarnett/.rvm/gems/ruby-2.5.1@metasploit-framework/bin/thin:23:in `<main>'
	from /home/jbarnett/.rvm/gems/ruby-2.5.1@metasploit-framework/bin/ruby_executable_hooks:15:in `eval'
	from /home/jbarnett/.rvm/gems/ruby-2.5.1@metasploit-framework/bin/ruby_executable_hooks:15:in `<main>'
```